python/pyarrow/meson.build

```diff
@@ -0,0 +1,426 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+py = import('python').find_installation(pure: false)
+
+# When NumPy 2.0 becomes the minimum we can remove the
+# custom location check
+numpy_dep = dependency('numpy', required: false)
```

Comment:
Currently numpy is a required build dependency but an optional runtime dependency. This seems to point we can build without numpy, is that the case?

Reply:
I haven't explored in detail, but a quick grep of the python code base shows the following includes of NumPy:

```sh
./python/pyarrow/src/arrow/python/helpers.h:29:#include <numpy/halffloat.h>
./python/pyarrow/src/arrow/python/type_traits.h:29:#include <numpy/halffloat.h>
./python/pyarrow/src/arrow/python/numpy_interop.h:22:#include <numpy/numpyconfig.h>  // IWYU pragma: export
./python/pyarrow/src/arrow/python/numpy_interop.h:41:#include <numpy/arrayobject.h>   // IWYU pragma: export
./python/pyarrow/src/arrow/python/numpy_interop.h:42:#include <numpy/arrayscalars.h>  // IWYU pragma: export
./python/pyarrow/src/arrow/python/numpy_interop.h:43:#include <numpy/ufuncobject.h>   // IWYU pragma: export
```

Reply:
yes, what I meant is that, shouldn't required be true?

Reply:
Ah I see. The comment directly preceding this in the code touches on this but maybe needs clarification. When NumPy 2.0 becomes the minimum yes you can make that check required and it will work as you expect, as that support is built into Meson. However, since we still need to support NumPy < 2, this injects some custom code to fall back to calling `np.get_include()` whenever the system cannot detect NumPy

Reply:
See also https://github.com/mesonbuild/meson/pull/12891 (ping @rgommers)